src/org/rascalmpl/library/lang/rascal/format/Rascal.rsc

```diff
@@ -0,0 +1,170 @@
+}
+@synopsis{Composes a default formatter for Rascal modules}
+@description{
+This module composes and describes a "standard" formatting style for Rascal.
+There could be other styles of course. Other styles can be build by 
+writing different `toBox` rules.
+}
+@bootstrapParser
+module lang::rascal::format::Rascal
+ 
+// by extending these modules we compose a `toBox` function
+// which handles all relevant constructs of Rascal
+extend lang::box::util::Tree2Box;
+extend lang::rascal::\syntax::Rascal;
+
+import ParseTree;
+import analysis::diff::edits::ExecuteTextEdits;
+import analysis::diff::edits::HiFiLayoutDiff;
+import analysis::diff::edits::TextEdits;
+import lang::box::\syntax::Box;
+import lang::box::util::Box2Text;
+import String;
+
+@synopsis{Format an entire Rascal file, in-place.}
+void formatRascalFile(loc \module) {
+    start[Module] tree = parse(#start[Module], \module);
+    edits = formatRascalModule(tree);
+    executeFileSystemChanges(changed(\module, edits));
+}
+
+@synopsis{Format a Rascal module string}
+str formatRascalString(str \module) 
+    = executeTextEdits(\module, formatRascalModule(parse(#start[Module], \module, |unknown:///|)));
+
+@synopsis{Top-level work-horse for formatting Rascal modules}
+@benefits{
+* retains source code comments 
+* uses Box for adaptive nested formatting
+}
+list[TextEdit] formatRascalModule(start[Module] \module) 
+    = layoutDiff(\module, parse(#start[Module], format(toBox(\module)), \module@\loc.top));
+
+
+/* Modules */
+
+Box toBox(Toplevel* toplevels) = V([toBox(t) | t <- toplevels], vs=2);
+
+Box toBox((Module) `<Tags tags> module <QualifiedName name> <Import* imports> <Body body>`)
+    = V([
+        toBox(tags),
+        H([L("module"), toBox(name)]),
+        toBox(imports),
+        toBox(body)
+    ], vs=1);
+
+Box toBox(Import* imports) = V([toBox(i) | i <- imports]);
+
+Box toBox((Import) `import <ImportedModule m> ;`)
+    = H([L("import"), H([toBox(m), L(";")], hs=0)]);
+
+Box toBox((Visibility) ``) = NULL();
+
+/* Declarations */
+
+Box toBox(FunctionModifier* modifiers) = H([toBox(b) | b <- modifiers]);
+
+Box toBox((Signature) `<FunctionModifiers modifiers> <Type typ>  <Name name> <Parameters parameters> throws <{Type ","}+ exs>`)
+    = HOV([
+        H([toBox(modifiers), toBox(typ), H([toBox(name), toBox(parameters)], hs=0)]), 
+        H([L("throws"), HV([toBox(e) | e <- exs])])], hs=1);
```

Comment:
Lists (like `exs` here) are a bit confusing, since iterating over them throws away the separators. In this case, we'll introduce a parse error where `throws` lists more than one `Type`:
```rascal
rascal>lexical A = "a";

ok
rascal>syntax AA = {A ","}* as;
ok                                                                                                                                                                                                                                                                            
rascal>aaa = [AA] "a,a,a";
AA: (AA) `a,a,a`
rascal>["<a>" | a <- aaa.as];
list[str]: ["a","a","a"]
```

Reply:
I can imagine this motivated added the `SL` operator. Nice 👍

Reply:
Right. List iteration which ignores separators is handy in some sense, but also confusing.